oak_core/src/sync.rs

```diff
@@ -50,6 +50,16 @@ impl<T> OnceCell<T> {
         Some(unsafe { (*self.value.get()).assume_init_ref() })
     }
 
+    pub fn get_mut(&self) -> Option<&mut T> {
```

Comment:
Breadcrumbs for myself for tomorrow: this fn is a copy-paste of `get()` above but it's dangerous (no matter the copy-pasted safety argument below) as it can leak mutable references.
I'll either have to flip it around (something like `mutate(fn: FnOnce(&mut T)` that calls `fn` while holding the lock) or just mark this as `unsafe` and let the callers deal with sorting out the guarantees.

Reply:
I don't think this will ever be safe, even it you use `mutate` or leave it up to the caller. One option is to require a `&mut self`, but I don't think that will work for your purposes. The other option is to have an atomic reference counter to be sure there are no other current references so taking a mutable reference becomes safe, but that basically just reimplements an atomic refcell.

For the case of the serial port, we will only ever need mutable references to it, so I would suggest using a `OnceCell<Spnlock<SerialPort>>` and lock it for every write to ensure safe mutable access.

Reply:
Actually, since the serial port will only ever be useful when a mutable reference is available the OnceCell does not really add value. You could just use `SpinLock<Option<SerialPort>>`.